File: quant1x/level1/security_list.rs
use super::sequence_id;
use crate::level1::int_to_float64;
use crate::level1::protocol::{self, commands, Request, RequestHeader, Response, ResponseHeader};
use crate::std::BinaryStream;
use encoding_rs::GBK;

// Request builder for SECURITY_LIST
#[allow(dead_code)]
#[derive(Debug, Clone)]
pub struct SecurityListRequest {
    header: RequestHeader,
    market: u16,
    start: u32,
    count: u32,
    unknown: u32,
}

impl SecurityListRequest {
    pub fn new(market: u16, start: u32, count: u32) -> Self {
        let mut header = RequestHeader::new();
        header.zip_flag = 0x0C;
        header.seq_id = sequence_id();
        header.packet_type = 0x01;
        header.method = commands::SECURITY_LIST;

        SecurityListRequest {
            header,
            market,
            start,
            count,
            unknown: 0,
        }
    }

    pub fn market(&self) -> u16 {
        self.market
    }

    pub fn start(&self) -> u32 {
        self.start
    }
}

impl Request for SecurityListRequest {
    fn header(&self) -> &RequestHeader {
        &self.header
    }

    fn header_mut(&mut self) -> &mut RequestHeader {
        &mut self.header
    }

    fn serialize_payload(&mut self) -> Vec<u8> {
        let mut payload = BinaryStream::new();
        payload.push_u16(self.market);
        payload.push_u32(self.start);
        payload.push_u32(self.count);
        payload.push_u32(self.unknown);
        payload.data().clone()
    }

    fn payload_string(&self) -> String {
        format!("{{Market:{}, Start:{}}}", self.market, self.start)
    }
}

#[allow(dead_code)]
#[derive(Debug, Clone)]
pub struct Security {
    pub code: String,
    pub vol_unit: u16,
    pub decimal_point: u8,
    pub name: String,
    pub pre_close: f64,
}

#[allow(dead_code)]
#[derive(Debug, Clone)]
pub struct SecurityListResponse {
    header: ResponseHeader,
    pub count: u16,
    pub list: Vec<Security>,
}

#[allow(dead_code)]
impl SecurityListResponse {
    pub fn new() -> Self {
        Self {
            header: ResponseHeader::new(),
            count: 0,
            list: Vec::new(),
        }
    }
}

impl Response for SecurityListResponse {
    fn header(&self) -> &ResponseHeader {
        &self.header
    }

    fn header_mut(&mut self) -> &mut ResponseHeader {
        &mut self.header
    }

    fn deserialize_body(&mut self, data: &[u8]) -> Result<(), crate::std::DeserializeError> {
        self.count = 0;
        self.list.clear();
        let mut bs = BinaryStream::from_vec(data.to_vec());
        if bs.data().len().saturating_sub(bs.position()) < 2 {
            return Ok(());
        }
        self.count = bs.get_u16()?;
        if self.count == 0 {
            return Ok(());
        }
        let min_required = 2 + (self.count as usize) * 25;
        if data.len() < min_required {
            log::warn!(
                "insufficient data for {} securities: data len {}, min required {}",
                self.count,
                data.len(),
                min_required
            );
            self.count = 0;
            return Ok(());
        }

        for _ in 0..self.count {
            let code = bs.get_string(6)?;
            let vol_unit = bs.get_u16()?;
            let mut name_buf = [0u8; 8];
            bs.get_byte_array(&mut name_buf)?;
            let name_nul = name_buf.iter().position(|&b| b == 0).unwrap_or(8);
            let (name_cow, _, _) = GBK.decode(&name_buf[..name_nul]);
            let name = name_cow.into_owned();
            let mut _rev1 = [0u8; 8];
            bs.get_byte_array(&mut _rev1)?;
            let mut _rev2 = [0u8; 4];
            bs.get_byte_array(&mut _rev2)?;
            let decimal_point = bs.get_u8()?;
            let tmp = bs.get_u32()?;
            let pre_close = int_to_float64(tmp);
            let mut _rev3 = [0u8; 4];
            bs.get_byte_array(&mut _rev3)?;

            self.list.push(Security {
                code,
                vol_unit,
                decimal_point,
                name,
                pre_close,
            });
        }
        Ok(())
    }

    fn body_string(&self) -> String {
        format!("{{Count:{}, Parsed:{}}}", self.count, self.list.len())
    }
}

/// Fetch a single page of security list from level1 server.
/// Returns Some(SecurityListResponse) on success, None on any IO error.
pub fn fetch_security_list(market: u16, start: u32, count: u32) -> Option<SecurityListResponse> {
    match crate::level1::client::client() {
        Ok(mut pooled) => {
            let mut request = SecurityListRequest::new(market, start, count);
            let mut response = SecurityListResponse::new();
            match protocol::process(pooled.stream(), &mut request, &mut response) {
                Ok(_) => {
                    log::info!(
                        "level1::security_list - market={} start={} count={}",
                        market,
                        start,
                        response.count
                    );
                    Some(response)
                }
                Err(e) => {
                    log::error!(
                        "level1 protocol::process error for security_list market={} start={}: {}",
                        market,
                        start,
                        e
                    );
                    None
                }
            }
        }
        Err(e) => {
            log::error!(
                "failed to acquire level1 client for security_list market={} start={}: {}",
                market,
                start,
                e
            );
            None
        }
    }
}

File: quant1x/level1/security_list.py
"""从 level1 服务器获取 SECURITY_LIST 页面（Python 实现）。

本模块提供一个小工具函数 `fetch_security_list(market, start, count)`，
其行为与 native（C++/Rust）实现保持一致：构造 SECURITY_LIST 请求，
通过连接池发送请求，并将响应体解析为字典列表返回。

注意：实现偏稳健与保守：遇到任意 I/O 或解析错误时返回 `None`，由调用方决定如何处理。
"""
from __future__ import annotations

import struct
import logging
from typing import Optional, List, Dict

from quant1x.level1 import client as l1client
from quant1x.level1 import protocol

log = logging.getLogger(__name__)

# 保持与 C++/Rust 实现类似的默认分页大小。C++ 头文件使用
# `security_list_pre_request_max`（通常为 1600），这里也使用 1600 作为默认值。
PRE_REQUEST_MAX = 1600


def _int_to_float64(v: int) -> float:
    # 移植自 Rust 中 level1/helpers.rs 的 `int_to_float64` 实现
    if v == 0:
        return 0.0
    log_point = ((v >> 24) & 0xFF)
    hleax = ((v >> 16) & 0xFF)
    lheax = ((v >> 8) & 0xFF)
    lleax = (v & 0xFF)

    dw_ecx = log_point * 2 - 0x7F
    dw_edx = log_point * 2 - 0x86
    dw_esi = log_point * 2 - 0x8E
    dw_eax = log_point * 2 - 0x96

    def pow2(i: int) -> float:
        return 2.0 ** i

    tmp_eax = -dw_ecx if dw_ecx < 0 else dw_ecx
    dbl_xmm6 = pow2(tmp_eax)
    if dw_ecx < 0:
        dbl_xmm6 = 1.0 / dbl_xmm6

    if hleax > 0x80:
        dwtmpeax = dw_edx + 1
        tmpdbl_xmm3 = pow2(dwtmpeax)
        dbl_xmm0 = pow2(dw_edx) * 128.0
        dbl_xmm0 += (hleax & 0x7F) * tmpdbl_xmm3
        dbl_xmm4 = dbl_xmm0
    elif dw_edx >= 0:
        dbl_xmm4 = pow2(dw_edx) * float(hleax)
    else:
        dbl_xmm4 = (1.0 / pow2(-dw_edx)) * float(hleax)

    dbl_xmm3 = pow2(dw_esi) * float(lheax)
    dbl_xmm1 = pow2(dw_eax) * float(lleax)

    if (hleax & 0x80) != 0:
        dbl_xmm3 *= 2.0
        dbl_xmm1 *= 2.0

    return dbl_xmm6 + dbl_xmm4 + dbl_xmm3 + dbl_xmm1


def fetch_security_list(market: int, start: int, count: int) -> Optional[List[Dict]]:
    """从 level1 服务器获取一页 SECURITY_LIST。

    返回一个字典列表，字典包含字段：`Code`（6 字符字符串）、`VolUnit`（整数）、
    `DecimalPoint`（整数）、`Name`（字符串）、`PreClose`（浮点）。出现错误时返回 `None`。
    """
    try:
        class SecurityListRequest:
            def __init__(self, market, start, count):
                self.market = market
                self.start = start
                self.count = count
            
            def serialize(self):
                payload = struct.pack('<H I I I', int(self.market) & 0xFFFF, int(self.start) & 0xFFFFFFFF, int(self.count) & 0xFFFFFFFF, 0)
                zip_flag = 0x0C
                seq_id = protocol.sequence_id()
                packet_type = 0x01
                pkg_len1 = 2 + len(payload)
                pkg_len2 = pkg_len1
                method = 0x044d
                header = struct.pack('<B I B H H H', zip_flag, seq_id, packet_type, pkg_len1, pkg_len2, method)
                return header + payload

        class SecurityListResponse:
            def __init__(self):
                self.body = b''
            
            def deserialize(self, data):
                self.body = data

        req = SecurityListRequest(market, start, count)
        resp = SecurityListResponse()

        with l1client.get_std_conn() as conn:
            protocol.process(conn, req, resp)
        
        body = resp.body

        if not body:
            # 响应体为空 -> 表示没有证券记录
            return []

        # 解析：先读取 u16 的计数，然后依次解析记录
        offset = 0
        if len(body) < 2:
            return []
        (cnt,) = struct.unpack_from('<H', body, offset)
        offset += 2
        result = []
        # 每条记录至少为 25 字节（与 Rust 实现一致）
        for _ in range(cnt):
            if offset + 25 > len(body):
                log.warning('Insufficient data when parsing SECURITY_LIST payload')
                break
            code_bytes = body[offset:offset+6]
            offset += 6
            (vol_unit,) = struct.unpack_from('<H', body, offset)
            offset += 2
            name_buf = body[offset:offset+16]
            offset += 16
            # 跳过 4 字节（保留字段）
            offset += 4
            (decimal_point,) = struct.unpack_from('<B', body, offset)
            offset += 1
            (tmp_u32,) = struct.unpack_from('<I', body, offset)
            offset += 4
            # 跳过最后 4 字节（保留/未使用）
            offset += 4

            # 解码代码和名称字段
            try:
                code = code_bytes.decode('ascii', errors='ignore').rstrip('\x00')
            except Exception:
                code = code_bytes.decode('utf-8', errors='ignore').rstrip('\x00')
            # 名称使用 GBK 编码，直到第一个 NUL 字节为止
            try:
                nul_pos = name_buf.index(0)
            except ValueError:
                nul_pos = len(name_buf)
            try:
                name = name_buf[:nul_pos].decode('gbk', errors='ignore')
            except Exception:
                name = name_buf[:nul_pos].decode('utf-8', errors='ignore')

            pre_close = _int_to_float64(tmp_u32)

            result.append({
                'Code': code,
                'VolUnit': int(vol_unit),
                'DecimalPoint': int(decimal_point),
                'Name': name,
                'PreClose': pre_close,
            })

        log.info('security_list fetched market=%s start=%s count=%s parsed=%s', market, start, count, len(result))
        return result
    except Exception as e:
        log.exception('fetch_security_list failed: %s', e)
        return None


if __name__ == '__main__':
    list = fetch_security_list(1, 0, 10)
    print(list)